### DISCUSSION TITLE
[Go] Trouble getting the resolved type of an argument from nested function

### DISCUSSION THREAD

**Discussion by rh-tguittet** [github/codeql#15886] [Q&A] [answered]
Hi, I am facing an issue to obtain the resolved type of an argument in the following situation (simplified from what some flavour of the Kubernetes client-go do) :

```go
package main

type Object interface{ Bar() bool }

type SpecificObject struct {
	Foo      string
	otherFoo string
}

func (so SpecificObject) Bar() bool { return true }

func cacheReader(key string) interface{} {
	if len(key) > 0 {
		// The details don't matter so much here, what counts is that an interface{} is returned
		return new(interface{})
	}

	return nil
}

func ReadObject(key string, obj Object) {
	obj = cacheReader(key).(Object)
}

func ReadObjectWrapper(obj Object) {
	ReadObject("foobar", obj)
}

func ReadObjectAfterCheck(obj Object) {
	if obj.Bar() { // The nature of the check does not matter here
		ReadObjectWrapper(obj)
	}
}

func sink(s string) { /**/ }

func main() {
	so := &SpecificObject{Foo: "Bar"}
	ReadObjectAfterCheck(so)
	sink(so.otherFoo)
}
```

With the following query:
```codeql
/**
 * @kind path-problem
 */

import go
import DataFlow

class MySource extends Function {
  MySource() { this.getName() = "ReadObject" }

  FunctionInput getInput() { result.isParameter(1) }

  FunctionOutput getOutput() { result.isParameter(1) }
}

private module TestFlow implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) {
    exists(MySource ms | node = ms.getOutput().getNode(ms.getACall()))
  }

  predicate isSink(DataFlow::Node node) {
    exists(Function sink | sink.getName() = "sink" and sink.getACall().getAnArgument() = node)
  }
}

module Flow = TaintTracking::Global<TestFlow>;

import Flow::PathGraph

from Flow::PathNode source, Flow::PathNode sink, Type t
where Flow::flowPath(source, sink) and t = source.getNode().getType()
select source.getNode(), source, sink, "$@ (with type $@) -> $@", source.getNode(), "__source__", t,
  t.pp(), sink.getNode(), "__sink__"
```

This returns: `__source__ (with type Object) -> __sink__`
Whereas I want `Object` to be a `PointerType` to `SpecificObject`. The CodeQL query works fine as long as there isn't nested calls to "parent types" (e.g. `ReadObjectAfterCheck>ReadObjectWrapper>ReadObject`)

It's not clear to me if the expectation I have are misguided or if some additional bits (like a flow to model `StructLit ~> Argument`) are required in the query. Please advise :)

**Comment by owen-mc** [comment]
Thank you for including a minimized example and a working query, which made it very easy to understand your question. The problem is that your source, which is `obj` in `ReadObjectWrapper`, has type `Object`. If you want it to give you the type from `so` in `main`, then you will have to make `so` the source. We want to find paths A -> B -> C, where B -> C is what you are currently finding and A is the definition of the variable in the first place. This can be done using flow states. I suggest the following approach:

- Use data flow with state (implement `DataFlow::StateConfigSig` and use `TaintTracking::GlobalWithState`) where the state is a newtype with two values ([example here](https://github.com/github/codeql/blob/820c14577a9146175d792ee5111d2297050c50c1/java/ql/lib/semmle/code/java/security/ArbitraryApkInstallationQuery.qll#L30-L32)), called something like `InitialState` and `HasBeenPassedToReadObject`.
- Make your source something like the initial definition of a variable (do you know it will always be a struct?) and always give it the state `InitialState`.
- Have an additional flow step where `state1` is `InitialState`, `state2` is `HasBeenPassedToReadObject`, with `node1` being the argument to `ReadObject` and `node2` being the post-update node of `node1` (which represents the value of `node` after the function call has happened). (This replaces the source definition in your current query.)

I hope that this approach works for you. I have been very vague on exactly how to define the source because I am not sure and it depends a bit on the exact details of your situation, so do try some different things and ask more questions if you need more help with that bit.

**Reply by rh-tguittet** [reply]
Thank you Owen for the prompt response.

Yes, my source will always be a pointer to a `StructLit` (`cacheReader()` essentially fills the object).

I had some issues with the 3rd point in your comment: I must be using the wrong expression to obtain the `PostUpdateNode` for `node2`. Here's what I got so far:

```codeql
/**
 * @kind path-problem
 */

import go
import DataFlow

class MySource extends Function {
  MySource() { this.getName() = "ReadObject" }
  FunctionInput getInput() { result.isParameter(1) }
  FunctionOutput getOutput() { result.isParameter(1) }
}

private newtype State =
  InitialState() or
  HasBeenPassedToReadObject()

private module TestFlow implements DataFlow::StateConfigSig {
  class FlowState = State;

  predicate isSource(DataFlow::Node node, FlowState state) {
    node.asExpr() instanceof StructLit and
    state instanceof InitialState
  }

  predicate isSink(DataFlow::Node node, FlowState state) {
    exists(Function sink | sink.getName() = "sink" and sink.getACall().getAnArgument() = node) and
    state instanceof HasBeenPassedToReadObject
  }

  predicate isAdditionalFlowStep(
    DataFlow::Node nodeFrom, FlowState stateFrom, DataFlow::Node nodeTo, FlowState stateTo
  ) {
    stateFrom instanceof InitialState and
    stateTo instanceof HasBeenPassedToReadObject and
    exists(MySource ms, CallNode call |
      call = ms.getACall() and
      nodeFrom = ms.getInput().getEntryNode(call).(ArgumentNode) and
      // I know this is awful but... the commented bit doesn't work
      nodeTo instanceof PostUpdateNode // and nodeTo = ms.getOutput().getExitNode(call)
    )
  }
}

module Flow = TaintTracking::GlobalWithState<TestFlow>;
import Flow::PathGraph

from Flow::PathNode source, Flow::PathNode sink, Type t
where
  Flow::flowPath(source, sink) and
  t = source.getNode().getType()
select sink.getNode(), source, sink, "$@ (with type $@) -> $@", source.getNode(),
  source.getNode().toString(), t, t.pp(), sink.getNode(), sink.getNode().toString()
```

To be fair it returns the results but when I look closely at the path the `nodeTo instanceof PostUpdateNode` is clearly off. How would you get the right `nodeTo` ?

**Reply by owen-mc** [reply]
Okay, there are two issues to discuss. First, what that line should be. Second, why are there no results.

Actually my point 3 was a bit misleading, because [OutParameter.getExitNode](https://github.com/github/codeql/blob/820c14577a9146175d792ee5111d2297050c50c1/go/ql/lib/semmle/go/dataflow/FunctionInputsAndOutputs.qll#L301-L312) already deals with this, so you could do this and not have to think about post-update nodes at all (if you quick-evaluate this, ignore the fact that the post-update node is in a weird place - this is a quirk of the way we currently do post-update nodes in go):
```
nodeTo = ms.getOutput().getExitNode(call)
```
The way to do it with post-update nodes is this:
```
      nodeTo.(DataFlow::PostUpdateNode).getPreUpdateNode() = nodeFrom
```
This means that you don't need `getOutput` - if it is always the same as `getInput` then that might be an advantage.

As for the second problem: I didn't notice before that there isn't actually flow from `obj` as an argument to `ReadObject` to `so` as an argument to `sink`. Any changes to `obj` that are made in `ReadObject` are lost because `obj` doesn't escape `ReadObjectWrapper`. If that's what you intended then the problem isn't "find flow A -> B -> C" but rather find flow "A -> B" and "A -> C", which is easily done with two dataflow configs.

**Reply by rh-tguittet** [reply]
> The way to do it with post-update nodes is this:
> ```codeql
> nodeTo.(DataFlow::PostUpdateNode).getPreUpdateNode() = nodeFrom
> ```

I did try that variant, it didn't work.

> Any changes to `obj` that are made in `ReadObject` are lost because `obj` doesn't escape `ReadObjectWrapper`

That is true, nothing is returned, `obj` is however updated. Can we consider this to be a result somehow?

> If that's what you intended then the problem isn't "find flow A -> B -> C" but rather find flow "A -> B" and "A -> C", which is easily done with two dataflow configs.

I didn't think of this one, but I did try "A -> B and B -> C". Modelling "A -> B and A -> C" doesn't return a valid result (only the shortest path "A -> C" is returned, we never visit "B").

Something I don't understand with "A -> B and A -> C" is how you can have "A -> C" take/show some of the path "A -> B" took. Won't we end up with diverging paths if we treat those paths without ensuring significant overlap?

**Reply by owen-mc** [reply]
Can you give me more detail about how `obj` in `ReadObject` being updated affects the rest of the program? Since it is a copy of a struct, and is discarded when the function returns,  this doesn't have any consequence on the rest of the program. I think that some important detail of the real situation has been lost in making this reduced example. I wonder if perhaps it should be the pointer-to-struct type `*SpecificObject` that implements the interface `Object`, rather than the struct type `SpecificObject`. Is it that the value of `so.otherFoo` can be somehow updated in `ReadObject`?

**Reply by rh-tguittet** [reply]
Right, so I changed `ReadObject` to this:
```go
func ReadObject(key string, obj Object) {
	//reflect.ValueOf(obj).Set(reflect.ValueOf(cacheReader(key).(Object)))
	obj.(*SpecificObject).otherFoo = "test"
}
```

But even with this, I cannot get the tainttracking with FlowState to work. "A->B and B->C" kinda work but I don't know how to stitch together the two flows (tried with `MergePathGraph` mirroring CommandInjection.ql but got some undecipherable error: "ERROR: DataFlow::DataFlowMake::MergePathGraph::TPathNode1 incompatible with DataFlow::DataFlowMake::MergePathGraph::TPathNode2 in context").

**Reply by owen-mc** [reply]
There appears to be a problem in the CodeQL dataflow analysis. `nodeTo = ms.getOutput().getExitNode(call)` is definitely correct, and gives the right result when I quick-evaluate it. The approach with flow states (A->B->C) should work. The problem is that the dataflow analysis doesn't realise there can be any flow out of the first parameter of `ReadObjectWrapper`. I will look into this further and try to fix it. Thank you for bringing this to our attention.

**Reply by owen-mc** [reply]
Okay, with some help from some colleagues I have got to the bottom of this. There is a heuristic in the dataflow library which doesn't allow flow into a function via a parameter and then back out of the same parameter. There is a way of turning this heuristic off in certain circumstances, but it is on a whole-language basis rather than being able to set it for a particular flow config. If you are able to edit your CodeQL library then you can add a disjunct to [this predicate](https://github.com/github/codeql/blob/b8608a1b1ce03c36d5fd2c6744f8ad39836d290d/go/ql/lib/semmle/go/dataflow/internal/DataFlowPrivate.qll#L420-L425) to make it hold for the particular case you are interested in. (Here are some examples from other languages: [C#](https://github.com/github/codeql/blob/b8608a1b1ce03c36d5fd2c6744f8ad39836d290d/csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll#L2882-L2890), [Java](https://github.com/github/codeql/blob/b8608a1b1ce03c36d5fd2c6744f8ad39836d290d/java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll#L586-L593), [Python](https://github.com/github/codeql/blob/b8608a1b1ce03c36d5fd2c6744f8ad39836d290d/python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll#L1067-L1077), [Ruby](https://github.com/github/codeql/blob/b8608a1b1ce03c36d5fd2c6744f8ad39836d290d/ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll#L2188-L2196), [Swift](https://github.com/github/codeql/blob/b8608a1b1ce03c36d5fd2c6744f8ad39836d290d/swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll#L1406-L1416).)

If you are not able to do that then I think the easiest approach is to have two separate configs, A->B and B->C. I see above that you get an error message from `MergePathGraph`. If you share the query with me then I can see what the problem is and fix it.

**Reply by rh-tguittet** [reply]
Thanks Owen for following up on this issue.

Changing the library on my laptop is easy... but it will be a massive headache if we choose to move forward with a more extensive PoC.

The query I am using for MergePathGraph:
```codeql
/**
 * @kind path-problem
 */
import go

module AtoB implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) { node.asExpr() instanceof StructLit }

  predicate isSink(DataFlow::Node node) {
    exists(Function f | f.getName() = "ReadObject" and node = f.getACall().getArgument(1))
  }
}

module BtoC implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) {
    exists(DataFlow::Node n2 |
      node.(DataFlow::PostUpdateNode).getPreUpdateNode() = n2 and AtoB::isSink(n2)
    )
  }

  predicate isSink(DataFlow::Node node) {
    exists(Function sink | sink.getName() = "sink" and sink.getACall().getAnArgument() = node)
  }
}

module FlowAtoB = TaintTracking::Global<AtoB>;

module FlowBtoC = TaintTracking::Global<BtoC>;

module FlowABC =
  DataFlow::MergePathGraph<FlowAtoB::PathNode, FlowBtoC::PathNode, FlowAtoB::PathGraph,
    FlowBtoC::PathGraph>;

import FlowABC::PathGraph

from FlowABC::PathNode source, FlowABC::PathNode sink, Type t
where
  t = source.getNode().getType() and
  FlowAtoB::flowPath(source.asPathNode1(), sink.asPathNode1()) and
  FlowBtoC::flowPath(source.asPathNode2(), sink.asPathNode2())
select sink.asPathNode2().getNode(), source, sink, "$@ (with type $@) -> $@", source.getNode(),
  source.getNode().toString(), t, t.pp(), sink.getNode(), sink.getNode().toString()
```

**Reply by owen-mc** [reply]
`MergePathGraph` doesn't do quite what you are trying to use it to do. (It doesn't concatenate paths.) In this case I think you don't need to show paths from both configs, so we don't need to deal with two different path graphs at all. This code worked for me.

```
/**
 * @kind path-problem
 */

import go

DataFlow::Node getArgumentToReadObject() {
  exists(Function f | f.getName() = "ReadObject" and result = f.getACall().getArgument(1))
}

module AtoB implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) { node.asExpr() instanceof StructLit }

  predicate neverSkip(DataFlow::Node n) { any() }

  predicate isSink(DataFlow::Node node) { node = getArgumentToReadObject() }
}

module BtoC implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) {
    node.(DataFlow::PostUpdateNode).getPreUpdateNode() = getArgumentToReadObject()
  }

  predicate isSink(DataFlow::Node node) {
    exists(Function sink | sink.getName() = "sink" and sink.getACall().getAnArgument() = node)
  }
}

module FlowAtoB = TaintTracking::Global<AtoB>;

module FlowBtoC = TaintTracking::Global<BtoC>;

import FlowBtoC::PathGraph

from DataFlow::Node a, DataFlow::Node b1, FlowBtoC::PathNode b2, FlowBtoC::PathNode c, Type t
where
  FlowAtoB::flow(a, b1) and
  t = a.getType() and
  b2.getNode().(DataFlow::PostUpdateNode).getPreUpdateNode() = b1 and
  FlowBtoC::flowPath(b2, c)
select c.getNode(), b2, c, "$@ (initialised $@ with type $@) -> $@", b1, b1.toString(), a, "here",
  t, t.pp(), c.getNode(), c.getNode().toString()
```

**Reply by rh-tguittet** [reply]
OK, let's do it this way then. I assume there is no easy way of adding a step to the path?

**Reply by rh-tguittet** [reply]
Sorry, had to unmark this as the answer: it doesn't work as expected when there are 2 types flowing through `ReadObject` but only one instance goes to the sink.

![2024-03-22_18-32](https://github.com/github/codeql/assets/88336850/c9e10f3b-b4b4-4140-a804-c07ee5bca44a)


New code, adds an `OtherObject` type compatible with `Object`:
```go
package main

import (
	"fmt"
)

type Object interface{ Bar() bool }

type SpecificObject struct {
	Foo      string
	otherFoo string
}

type OtherObject struct {
	otherFoo string
}

func (so SpecificObject) Bar() bool { return true }
func (oo OtherObject) Bar() bool    { return true }

func ReadObject(key string, obj Object) {
	switch obj.(type) {
	case *SpecificObject:
		obj.(*SpecificObject).otherFoo = "bar"
	case *OtherObject:
		obj.(*OtherObject).otherFoo = "baz"
	}
}

func ReadObjectWrapper(obj Object) {
	ReadObject("foobar", obj)
}

func ReadObjectAfterCheck(obj Object) {
	if obj.Bar() { // The nature of the check does not matter here
		ReadObjectWrapper(obj)
	}
}

func sink(s string) { /**/ }

func main() {
	so := &SpecificObject{Foo: "Bar"}
	ReadObjectAfterCheck(so)
	fmt.Printf("so: %+v\n", so)
	oo := &OtherObject{}
	ReadObjectAfterCheck(oo)
	fmt.Printf("oo: %+v\n", oo)
	sink(so.otherFoo)
}
```

**Reply by owen-mc** [reply, marked as answer]
Good point. Try this:
```
/**
 * @kind path-problem
 */

import go

DataFlow::Node getA() { result.asExpr() instanceof StructLit }

DataFlow::Node getB() {
  exists(Function f | f.getName() = "ReadObject" and result = f.getACall().getArgument(1))
}

DataFlow::Node getC() {
  exists(Function sink | sink.getName() = "sink" and result = sink.getACall().getAnArgument())
}

module AtoB implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) { node = getA() }

  predicate isSink(DataFlow::Node node) { node = getB() }
}

module BtoC implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) {
    node.(DataFlow::PostUpdateNode).getPreUpdateNode() = getB()
  }

  predicate isSink(DataFlow::Node node) { node = getC() }
}

module AtoC implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node node) { node = getA() }

  predicate isSink(DataFlow::Node node) { node = getC() }
}

module FlowAtoB = TaintTracking::Global<AtoB>;

module FlowBtoC = TaintTracking::Global<BtoC>;

module FlowAtoC = TaintTracking::Global<AtoC>;

import FlowBtoC::PathGraph

from DataFlow::Node a, DataFlow::Node b1, FlowBtoC::PathNode b2, FlowBtoC::PathNode c, Type t
where
  FlowAtoB::flow(a, b1) and
  t = a.getType() and
  b2.getNode().(DataFlow::PostUpdateNode).getPreUpdateNode() = b1 and
  FlowBtoC::flowPath(b2, c) and
  FlowAtoC::flow(a, c.getNode())
select c.getNode(), b2, c, "$@ (initialised $@ with type $@) -> $@", b1, b1.toString(), a, "here",
  t, t.pp(), c.getNode(), c.getNode().toString()
```
It is a shame that using flow states did not work in your case. Perhaps I will see if I can get something added to allow over-riding that that predicate in the config. If I can I will report back here.

**Reply by rh-tguittet** [reply]
That worked. One note from adapting this to the non-example code base: `isAdditionalFlowStep` for A->C and B->C must be shared.

Thank you Owen for seeing it through!
